src/metabase/query_processor/middleware/add_implicit_joins.clj

```diff
@@ -30,63 +30,78 @@
-      (for [{fk-name :name, fk-field-id :id, fk-ident :ident, pk-id :fk-target-field-id} fk-fields
-            :when                                                                        pk-id]
+      (for [{fk-field-id :fk-field-id, fk-join-alias :fk-join-alias} fk-field-infos
+            :let [fk-field (lib.metadata.protocols/field (qp.store/metadata-provider) fk-field-id)]
+            :when fk-field
+            :let [{pk-id :fk-target-field-id, fk-name :name, fk-ident :ident} fk-field]
+            :when pk-id]
         (let [{source-table :table-id} (lib.metadata.protocols/field (qp.store/metadata-provider) pk-id)
               {table-name :name}       (lib.metadata.protocols/table (qp.store/metadata-provider) source-table)
-              alias-for-join           (join-alias table-name fk-name)]
-          (-> {:source-table source-table
-               :alias        alias-for-join
-               :ident        (lib/implicit-join-clause-ident fk-ident)
-               :fields       :none
-               :strategy     :left-join
-               :condition    [:= [:field fk-field-id nil] [:field pk-id {:join-alias alias-for-join}]]
-               :fk-field-id  fk-field-id}
+              alias-for-join           (join-alias table-name fk-name fk-join-alias)]
+          (-> (merge {:source-table  source-table
+                      :alias         alias-for-join
+                      :ident         (lib/implicit-join-clause-ident fk-ident)
+                      :fields        :none
+                      :strategy      :left-join
+                      :condition     [:= [:field fk-field-id (when fk-join-alias {:join-alias fk-join-alias})]
+                                      [:field pk-id {:join-alias alias-for-join}]]
+                      :fk-field-id   fk-field-id}
+                     (when fk-join-alias
+                       {:fk-join-alias fk-join-alias}))
               (vary-meta assoc ::needs [:field fk-field-id nil])))))))
 
-(defn- implicitly-joined-fields->joins
+(mu/defn- implicitly-joined-fields->joins :- [:maybe [:sequential JoinInfo]]
   "Create implicit join maps for a set of `field-clauses-with-source-field`."
   [field-clauses-with-source-field]
-  (distinct
-   (let [fk-field-ids (->> field-clauses-with-source-field
+  (let [k-field-infos (->> field-clauses-with-source-field
                            (map (fn [clause]
                                   (lib.util.match/match-one clause
                                     [:field (id :guard integer?) (opts :guard (every-pred :source-field (complement :join-alias)))]
-                                    (:source-field opts))))
-                           (filter integer?)
-                           set
+                                    {:fk-field-id (:source-field opts), :fk-join-alias (:source-field-join-alias opts)})))
+                           distinct
```

Comment:
This could be a `(sequence (comp (map ...) (distinct) field-clauses-with-source-field)` if we wanted to generate a bit less garbage. (Not important.)